subsys/random/rand32_timer.c

```diff
@@ -14,21 +14,38 @@
  * provide a random number generator.
  */
 
+#include <zephyr/init.h>
 #include <zephyr/random/rand32.h>
 #include <zephyr/drivers/timer/system_timer.h>
 #include <zephyr/kernel.h>
-#include <zephyr/sys/atomic.h>
 #include <string.h>
+#include <zephyr/device.h>
 
 #if defined(__GNUC__)
 
-/*
- * Symbols used to ensure a rapid series of calls to random number generator
- * return different values.
- */
-static atomic_val_t _rand32_counter;
+/* seed the state with pseudo-random data */
+static uint32_t state[4] = {0, 0x9b64c2b0, 0x86d3d2d4, 0xa00ae278};
+
+static inline uint32_t rotl(const uint32_t x, int k)
+{
+	return (x << k) | (x >> (32 - k));
+}
 
-#define _RAND32_INC 1000000003U
+static uint32_t xoshiro128_next(void)
+{
+	const uint32_t result = rotl(state[0] + state[3], 7) + state[0];
+	const uint32_t t = state[1] << 9;
+
+	state[2] ^= state[0];
+	state[3] ^= state[1];
+	state[1] ^= state[2];
+	state[0] ^= state[3];
+
+	state[2] ^= t;
+	state[3] = rotl(state[3], 11);
```

Comment:
I guess we want to add some locking here as we access to global data and we may call this from multiple treads. spinlock will suite perfectly fine.

Reply:
threads injecting here might actually help this test rng ;-)